### PR TITLE
Casual improvements

### PR DESCRIPTION
**Pull request by dkochmanski**
Supersedes #882. This branch is based on #886 so it should be reviewed after merging it.

### CODE REVIEW

File: Core/clim-basic/extended-streams/text-formatting.lisp [hunk 1]
```diff
@@ -44,7 +44,9 @@
                   (setf (getf text-margins ,edge) ,default))))
```
Comment (scymtym) [inline Core/clim-basic/extended-streams/text-formatting.lisp]: Commit message: "stream-text-margin" -> "stream-text-margins"

File: Core/clim-basic/extended-streams/text-formatting.lisp [hunk 2]
```diff
@@ -44,7 +44,9 @@
...
     (thunk :left   '(:absolute 0))
     (thunk :top    '(:absolute 0))
-    (thunk :right  (or text-margin '(:relative 0)))
+    (thunk :right  (if text-margin
+                       `(:absolute ,text-margin)
+                       `(:relative 0)))
     (thunk :bottom '(:relative 0))
```
Comment (scymtym) [inline Core/clim-basic/extended-streams/text-formatting.lisp]: I suggest extracting this logic into a function `(normalize-text-margins text-margins text-margin)` or similar and adding a unit test.

File: Core/clim-core/bordered-output.lisp [hunk 3]
```diff
@@ -456,8 +456,29 @@
                           (draw-line* stream left bottom right bottom
                                       :ink ink
                                       :line-style line-style)))
-                       (updating-output-record  nil)
-                       (compound-output-record  (fn child))))))
+                       (updating-output-record nil)
+                       (compound-output-record (fn child))))))
+      (fn record))))
+
+(define-border-type :crossout (stream record
```
Comment (scymtym) [inline Core/clim-core/bordered-output.lisp]: This is often called "strikethrough". Is there a specific reason for going with "crossout"?

File: Core/clim-core/bordered-output.lisp [hunk 4]
```diff
@@ -456,8 +456,29 @@
                           (draw-line* stream left bottom right bottom
```
Comment (scymtym) [inline Core/clim-core/bordered-output.lisp]: In `examples`, there is a bordered output test and a dedicated underlining test (within drawing tests?). One or both should test this feature.

File: Core/clim-basic/extended-streams/text-formatting.lisp [hunk 5]
```diff
@@ -22,7 +22,7 @@
   (ignore-errors ; destructuring-bind may error; that yields invalid spec
     (destructuring-bind (&key left top right bottom) margins
       (flet ((margin-spec-p (m)
-               (and (member (first m) '(:absolute :relative))
+               (and (member (first m) '(:relative :absolute))
```
Comment (scymtym) [inline Core/clim-basic/extended-streams/text-formatting.lisp]: Can you add a `destructuring-bind` to check the shape of `m` and document the meaning of its parts?

File: Core/clim-basic/extended-streams/text-formatting.lisp [hunk 6]
```diff
@@ -93,9 +95,10 @@
     (new-margins (stream standard-page-layout)
      &aux (old-margins (stream-text-margins stream)))
   (macrolet ((thunk (edge)
-               `(unless (getf new-margins ,edge)
-                  (setf (getf new-margins ,edge)
-                        (getf old-margins ,edge)))))
+               `(let ((spec (getf new-margins ,edge)))
+                  (typecase spec
+                    (null (setf (getf new-margins ,edge) (getf old-margins ,edge)))
```
Comment (scymtym) [inline Core/clim-basic/extended-streams/text-formatting.lisp]: Pull the common `setf (getf new-margins ,edge)` part out of the `typecase`.

File: Core/clim-basic/extended-streams/text-formatting.lisp [hunk 7]
```diff
@@ -31,23 +31,25 @@
...
 
 (defmethod initialize-instance :after ((instance standard-page-layout)
                                        &key text-margins text-margin)
   (macrolet ((thunk (edge default)
-               `(unless (getf text-margins ,edge)
-                  (setf (getf text-margins ,edge) ,default))))
-    (thunk :left   '(:absolute 0))
-    (thunk :top    '(:absolute 0))
+               `(let ((spec (getf text-margins ,edge)))
+                  (typecase spec
```
Comment (scymtym) [inline Core/clim-basic/extended-streams/text-formatting.lisp]: Since this is repeated below, add a `normalize-margin-spec` or similar helper function.

File: Core/clim-basic/extended-streams/text-formatting.lisp [hunk 8]
```diff
@@ -31,23 +31,25 @@
 
 (defclass standard-page-layout ()
   ((%page-region :reader stream-page-region :writer (setf %page-region))
-   (margins :initform '(:left   (:absolute 0)
-                        :top    (:absolute 0)
+   (margins :initform '(:left   (:relative 0)
+                        :top    (:relative 0)
```
Comment (scymtym) [inline Core/clim-basic/extended-streams/text-formatting.lisp]: Why is this changed? This does not seem purely syntactic.

File: Core/clim-basic/extended-streams/text-formatting.lisp [hunk 9]
```diff
@@ -50,7 +50,7 @@
                        `(:absolute ,text-margin)
                        `(:relative 0)))
     (thunk :bottom `(:relative 0))
-    (setf (stream-text-margins instance) text-margins)))
+    (setf (slot-value instance 'margins) text-margins)))
```
Comment (scymtym) [inline Core/clim-basic/extended-streams/text-formatting.lisp]: Why not use the accessor?

File: Core/clim-core/bordered-output.lisp (hunk 3 above)
Comment (dkochmanski) [reply Core/clim-core/bordered-output.lisp]: I've checked names and I had two choices with the same meaning: `crossout` and `strikethrough`, I've picked the shorter one.

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 8 above)
Comment (dkochmanski) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: left/top is always located at 0/0, I've changed it to `:relative` to match bottom/right defaults (so the person who reads the code doesn't have to wonder why they are different).

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 1 above)
Comment (dkochmanski) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: this commit message is correct, it addresses #880 where `:text-margin` - which is the initarg for a slot accessed with `stream-text-margin`, didn't work quite correct with standard-page-layout (and stream-text-margins).

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 6 above)
Comment (dkochmanski) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: it doesn't change the value when spec is a list which is not empty. pulling it out would be incorrect.

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 7 above)
Comment (dkochmanski) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: it is slightly different (below macrolet merges old and new specs while this one provides defaults)

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 6 above)
Comment (scymtym) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: I thought with a `normalize-margin-spec` function (mentioned in another comment below) which would be the identity on (well-formed) conses, this could be `(setf (getf text-margins ,edge) (normalize-margin-spec (getf text-margins ,edge)))`.

Now that I think about it some more, why not make a function that normalizes (and checks?) a whole "margin spec plist"? It could be used here as well as in `(setf stream-text-margins) :around` and also tested.

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 9 above)
Comment (dkochmanski) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: because `(setf stream-text-margins)` has an around method which tries to merge new margins with old ones (and that calls `stream-text-margins` on an the yet-unbound slot)

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 7 above)
Comment (scymtym) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: As far as I can tell, a function `normalize-margin-spec-plist plist defaults` called as

* `(normalize-margin-spec-plist text-margins '(:left (:absolute 0) :top (:absolute 0) …))`

* `(normalize-margin-spec-plist text-margins old-margins)`

would do the job.

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 2 above)
Comment (dkochmanski) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: this logic with default value is used only in this particular method (there is something similar but not quite the same in setf stream-text-margins) -- is there much of a value in extracting it?

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 5 above)
Comment (dkochmanski) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: given it looks like this:
```
(defun valid-margin-spec-p (margins)
  (ignore-errors ; destructuring-bind may error; that yields invalid spec
   (destructuring-bind (&key left top right bottom) margins
     (flet ((margin-spec-p (margin)
              (destructuring-bind (anchor value) margin
                (and (member anchor '(:relative :absolute))
                     (realp value)))))
       (every #'margin-spec-p (list left top right bottom))))))
```

how would you like me to document it?

(notice, that I've changed (not (null value)) to (realp value))

Comment (scymtym) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: `destructuring-bind (anchor value) margin` is exactly what I meant by documenting the shape using `destructuring-bind`.

File: Core/clim-core/bordered-output.lisp (hunk 4 above)
Comment (dkochmanski) [reply Core/clim-core/bordered-output.lisp]: done

**Comment by dkochmanski** [comment]
I've modified valid-margin-spec-p and added examples for crossout. I've also rebased against master.

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 7 above)
Comment (dkochmanski) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: done

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 6 above)
Comment (dkochmanski) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: ditto

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 7 above)
Comment (dkochmanski) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: done

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 6 above)
Comment (dkochmanski) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: done

File: Tests/text-formatting.lisp [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+(test normalize-margin-spec
+  (flet ((equals (spec-1 spec-2)
+           (and (= (length spec-1) (length spec-2))
+                (loop for (edge value) on spec-1 by #'cddr
+                      unless (equal value (getf spec-2 edge))
+                        do
+                           (return nil)
+                      finally
+                         (return t)))))
```
Comment (scymtym) [inline Tests/text-formatting.lisp]: `always (equal value (getf spec-2 edge))`

File: Tests/text-formatting.lisp [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    (is (equals (climi::normalize-margin-spec nil
+                                              '(:left (:relative 0)
+                                                :top (:relative 0)
+                                                :right (:relative 0)
+                                                :bottom (:relative 0)))
+                '(:left (:relative 0)
+                  :top (:relative 0)
+                  :right (:relative 0)
+                  :bottom (:relative 0))))
```
Comment (scymtym) [inline Tests/text-formatting.lisp]: Fiveam wants the expected value as the first argument. Putting the expected value first generates the report
```
(climi::normalize-margin-spec …)

  evaluated to

(:top …)
```

instead of

```
  (:top …)

evaluated to

  (:top …)
```

File: Tests/text-formatting.lisp [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+           :foo)
+         default))
+      (signals error
+        (climi::normalize-margin-spec
+         `(,(5am:gen-one-element :left :right :top :bottom)
+           (:relative 1 2))
+         default))
+      (signals error
+        (climi::normalize-margin-spec `(:left* (:relative 42))
+                                      default)))))
```
Comment (scymtym) [inline Tests/text-formatting.lisp]: * The `5am` package prefix is not needed since we `:use` the package
* `gen-one-element` returns a generator function, so ``(,(gen-one-element …) …)` splices the function, not a random value
* `gen-one-element` and other generators are intended to be used in conjunction with `for-all`

I suggest testing the validation behavior like this:
```lisp
(defun gen-invalid-margin-spec-element ()
  (let ((anchor (gen-one-element :relative :absolute)))
   (lambda ()
     (case (random 3)
       (0 :foo)
       (1 (list (funcall anchor) :foo))
       (2 (list (funcall anchor) 1 2))))))

(defun gen-invalid-margin-spec ()
  (let ((edge (gen-one-element :left :right :top :bottom))
        (element (gen-invalid-margin-spec-element)))
    (lambda ()
      (list (funcall edge) (funcall element)))))

(test normalize-margin-spec.validation

  (for-all ((spec (gen-invalid-margin-spec)))
    (let ((default '(:left (:relative 0)
                     :top (:relative 0)
                     :right (:relative 0)
                     :bottom (:relative 0))))
      (signals error (climi::normalize-margin-spec spec default)))))
```

File: Core/clim-basic/extended-streams/text-formatting.lisp [hunk 13]
```diff
@@ -40,17 +53,13 @@
...
-                       `(:relative 0)))
-    (thunk :bottom `(:relative 0))
+  (let ((right-margin (if text-margin
+                          `(:absolute ,text-margin)
+                          `(:relative 0))))
+    (setf text-margins (normalize-margin-spec text-margins `(:left   (:relative 0)
+                                                             :top    (:relative 0)
+                                                             :right  ,right-margin
+                                                             :bottom (:relative 0))))
     (setf (slot-value instance 'margins) text-margins)))
```
Comment (scymtym) [inline Core/clim-basic/extended-streams/text-formatting.lisp]: You could just `(setf (slot-value instance 'margins) (normalize …))` now.

File: Tests/text-formatting.lisp (hunk 10 above)
Comment (dkochmanski) [reply Tests/text-formatting.lisp]: done

File: Tests/text-formatting.lisp (hunk 11 above)
Comment (dkochmanski) [reply Tests/text-formatting.lisp]: done

File: Tests/text-formatting.lisp (hunk 12 above)
Comment (dkochmanski) [reply Tests/text-formatting.lisp]: done

File: Core/clim-basic/extended-streams/text-formatting.lisp (hunk 13 above)
Comment (dkochmanski) [reply Core/clim-basic/extended-streams/text-formatting.lisp]: done
